bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -1796,43 +1858,111 @@ OperationResult PreProcessor::launchReqPreProcessing(const string &batchCid,
+  // insufficient), send with a buffer size that must fit.
+  for (size_t i{}; i < 2; ++i) {
+    // reduce internal overheads from buffer size for execution engine response
+    auto maxReplySize = resultBufferEntry.get().size - responseInternalOverhead_;
+    request = bftEngine::IRequestsHandler::ExecutionRequest{
+        clientId,
+        reqSeqNum,
+        reqCid,
+        PRE_PROCESS_FLAG,
+        preProcessReqMsg->requestLength(),
+        preProcessReqMsg->requestBuf(),
+        std::string(preProcessReqMsg->requestSignature(), preProcessReqMsg->requestSignatureLength()),
+        maxReplySize,
+        resultBufferEntry.get().buffer,
+        reqSeqNum,
+        reqOffsetInBatch,
+        preProcessReqMsg->result()};
+    requestsHandler_.preExecute(request, std::nullopt, reqCid, span);
+    preProcessResult = static_cast<OperationResult>(request.outExecutionStatus);
+    resultLen = request.outActualReplySize;
+
+    if (preProcessResult == OperationResult::SUCCESS) {
+      break;
+    }
     LOG_ERROR(logger(),
-              "Pre-execution failed" << KVLOG(
-                  clientId, reqSeqNum, batchCid, reqCid, reqOffsetInBatch, (uint32_t)preProcessResult, resultLen));
-  }
+              "Pre-execution failed" << KVLOG(clientId,
+                                              reqSeqNum,
+                                              batchCid,
+                                              reqCid,
+                                              reqOffsetInBatch,
+                                              (uint32_t)preProcessResult,
+                                              resultLen,
+                                              request.outActualReplySize,
+                                              request.outRequiredReplySize,
+                                              maxReplySize,
+                                              resultBufferEntry.get().buffer,
+                                              request.maxReplySize,
+                                              request.requestSize));
+    if (!bufferPool_ || (OperationResult::EXEC_DATA_TOO_LARGE != preProcessResult)) {
+      // do not retry pre-execute if not using a memory pool OR if error is not due to response buffer size
+      break;
+    }
+
+    // If we are here: using memory pool and the error is due to insufficient buffer size
+    auto maxSupportedBufferSize = subpoolsConfig_.back().bufferSize;
+    auto args = KVLOG(i,
+                      reqSeqNum,
+                      request.maxReplySize,
+                      request.outRequiredReplySize,
+                      request.outActualReplySize,
+                      maxSupportedBufferSize);
+    if (i == 0) {  // 1st attempt
+      if ((request.outRequiredReplySize <= request.maxReplySize) ||
+          (request.outRequiredReplySize > maxSupportedBufferSize)) {
```

Comment:
What is wrong with "request.outRequiredReplySize <= request.maxReplySize"?
Why we don't support such requests?